src/ripple/core/TimeKeeper.h

```diff
@@ -22,73 +22,98 @@
 
-        Close time is the time the network would agree that
-        a ledger closed, if a ledger closed right now.
+        This estimate is not directly visible to other servers over the
+        protocol, but it is possible for them to make an educated guess
+        if this server publishes proposals or validations.
 
-        The close time represents the notional "center"
-        of the network. Each server assumes its clock
-        is correct, and tries to pull the close time towards
-        its measure of network time.
+        @note The network time is adjusted for the "Ripple epoch" which
+              was arbitrarily defined as 2000-01-01T00:00:00Z by Arthur
+              Britto and David Schwartz during early development of the
+              code. No rationale has been provided for this curious but
+              unimportant choice.
     */
-    virtual time_point
-    closeTime() const = 0;
+    [[nodiscard]] time_point
+    now() const override
+    {
+        return adjust(std::chrono::system_clock::now());
+    }
 
-    /** Adjust the close time.
+    /** Returns the predicted close time, in network time.
 
-        This is called in response to received validations.
+        The predicted close time represents the notional "center" of the
+        network. Each server assumes that its clock is correct and tries
+        to pull the close time towards its measure of network time.
     */
-    virtual void
-    adjustCloseTime(std::chrono::duration<std::int32_t> amount) = 0;
+    [[nodiscard]] time_point
+    closeTime() const
+    {
+        return now() + closeOffset_.load();
+    }
 
     // This may return a negative value
-    virtual std::chrono::duration<std::int32_t>
-    nowOffset() const = 0;
-
-    // This may return a negative value
-    virtual std::chrono::duration<std::int32_t>
-    closeOffset() const = 0;
+    [[nodiscard]] std::chrono::seconds
+    closeOffset() const
+    {
+        return closeOffset_.load();
+    }
+
+    /** Adjust the close time, based on the network's view of time. */
+    std::chrono::seconds
+    adjustCloseTime(std::chrono::seconds by)
+    {
+        using namespace std::chrono_literals;
+
+        auto offset = closeOffset_.load();
+
+        // The close time adjustment is serialized externally to this
+        // code. The compare/exchange only serves as a check here and
+        // should not fail. If it does, it's safe to simply just skip
+        // the adjustment.
+        [[maybe_unused]] auto const success =
+            closeOffset_.compare_exchange_strong(offset, [by, offset]() {
+                // Ignore small offsets and push the close time
+                // towards our wall time.
+                if (by > 1s)
+                    return offset + ((by + 3s) / 4);
+
+                if (by < -1s)
+                    return offset + ((by - 3s) / 4);
+
+                return (offset * 3) / 4;
+            }());
+        assert(success);
```

Comment:
I don't really object, but doesn't asserting here mean you can unconditionally assign? If the assert fails, that means another thread changed the value between the time we read it and changed it. If we're asserting that never fails do we need the compare and exchange strong? Do we need the atomic?

Honestly, I think I'd keep the atomics - it makes the code thread safe, but I might remove the `assert`. What breaks if that `assert` fires?

Reply:
You're right; we could just unconditionally assign and nothing _really_ goes wrong. My intention here was to implement a "weak" form of the precondition that calling this function is _supposed_ to be externally serialized, but it's _extremely_ weak and subject to timing, instead of being a robust check.

I can just change this to be:

```
        closeOffset_ = [by, offset]() {
                // Ignore small offsets and push the close time
                // towards our wall time.
                if (by > 1s)
                    return offset + ((by + 3s) / 4);

                if (by < -1s)
                    return offset + ((by - 3s) / 4);

                return (offset * 3) / 4;
            }();
```